### ISSUE TITLE
SP8 touchscreen and pen support

### ISSUE THREAD

**Issue by luca-viggiani** [linux-surface/linux-surface#818]
Hi I'm sure that this has already been asked before (maybe several times), but I can't find where.
And I know there's some work ongoing about getting touchscreen and pen work on SP8.
So what's the status and where can I track it?
Thanks!

**Comment by qzed** [comment]
Main discussion for the touch host controller is going on here: https://github.com/linux-surface/intel-precise-touch/issues/8. That will provide single- and (via IPTSd) multi-touch support. Pen support still requires reverse-engineering the new data format: https://github.com/linux-surface/intel-precise-touch/issues/14.

**Comment by luca-viggiani** [comment]
Thank'you very much! So at the moment touchscreen is still unsupported as stated in the feature matrix, right?

**Comment by qzed** [comment]
Yes, but @quo already has a preliminary driver that we can hopefully integrated soon.

**Comment by BeatVids** [comment]
Great news, I noticed that SP7+ is now supported in the feature matrix. The only unsupported one is now just the SP8. I've tried setting up ITHC, but no luck. Anything we can do now to get it to work for SP8? Please let me know if there's anything I can do to contribute.

**Comment by j4ckgrey** [comment]
Hi there, I also own SP8, if there is anything I can do to contribute, or testing, please just ask.

**Comment by quo** [comment]
The ithc driver should work now for SP8. Pen may also work with the experimental iptsd.

**Comment by j4ckgrey** [comment]
@quo mate, you are a hero. Thanks a lot

**Comment by luca-viggiani** [comment]
> @luca-viggiani last night it was tested by another user, and confirmed in an issue i created in the original driver's post, was confirmed to work, multi touch as well using the IPTSD experimental after some edit was done by quo .... so yeah, I would I would go ahead and install it then start searching for 120Hz support xD

That's great thanks!

**Comment by gurrgur** [comment]
The discussion can be found here: https://github.com/quo/ithc-linux/issues/4

Also I'd like to add that using https://github.com/quo/iptsd pen input works as well on my SP8 with a Slim Pen 2 but there are still some issues with tilt and the eraser not yet being recognized as an eraser (it works just like a regular pen).

**Comment by j4ckgrey** [comment]
> The discussion can be found here: [quo/ithc-linux#4](https://github.com/quo/ithc-linux/issues/4)
> 
> Also I'd like to add that using https://github.com/quo/iptsd pen input works as well on my SP8 with a Slim Pen 2 but there are still some issues with tilt and the eraser not yet being recognized as an eraser (it works just like a regular pen).

did you have any luck bringing 120hz to the screen?

**Comment by gurrgur** [comment]
Running it at 120Hz with variable refresh rate enabled right now :)
I got it working by manually purging all 60Hz modes from the edid in a hex editor. In #825 I describe a slightly different method which enables 120Hz but loses VRR in the process.

This is the edid which has the manual edits:
[edid-sp8-120hz-vrr-mod.zip](https://github.com/linux-surface/linux-surface/files/8805318/edid-sp8-120hz-vrr-mod.

The method for setting up an edid override is still the same as shown in #825.

AFAIK VRR in combination with an Intel iGPU is currently only supported by Plasma Wayland and Sway. Unlike with amdgpu I believe there is no xorg config option which would enable it in an X11 session...

**Comment by j4ckgrey** [comment]
@quo i didnt get touch working at first so I did enter poll mode, and it worked only then in one a single touch mode (I already have Linux-Surface kernel installed), but I tried to install iptsd and I keep getting ERROR: Pkg-config binary for machine 1 not found. Giving up.

Is there something im doing wrong?

**Comment by quo** [comment]
@j4ckgrey I think you just have to install pkg-config. On Ubuntu you should be able to install it with `sudo apt install pkg-config`.

As for poll mode, did you get the "Blocked an interrupt request" error in dmesg? If so, it's best to use the `intremap=nosid` workaround from the readme (`sudo make set-nosid` should work on Ubuntu).

**Comment by j4ckgrey** [comment]
> @j4ckgrey I think you just have to install pkg-config. On Ubuntu you should be able to install it with `sudo apt install pkg-config`.
> 
> As for poll mode, did you get the "Blocked an interrupt request" error in dmesg? If so, it's best to use the `intremap=nosid` workaround from the readme (`sudo make set-nosid` should work on Ubuntu).

Not really, i got no error in dmesg, everything was just running smooth. And yet touch wont work. I still get an error while installing the iptsd. In the step sudo ./build/iptsd

ERROR: ../src/control.c:139: Failed to open /dev/ipts/0: No such file or directory
ERROR: ../src/main.c:106: Failed to start IPTS: No such file or directory

**Comment by quo** [comment]
Not sure why it wouldn't work without poll in that case... @BeatVids @gurrgur Did you need the nosid or poll workarounds?

I think you may have the wrong iptsd version, you need this one: https://github.com/quo/iptsd

**Comment by j4ckgrey** [comment]
> Not sure why it wouldn't work without poll in that case... @BeatVids @gurrgur Did you need the nosid or poll workarounds?
> 
> I think you may have the wrong iptsd version, you need this one: https://github.com/quo/iptsd

This is the iptsd I'm installing .... Im also confused cause it worked with others.

If @BeatVids @gurrgur you guys got any insight please tell, what distro are you using?

**Comment by gurrgur** [comment]
> ERROR: ../src/control.c:139: Failed to open /dev/ipts/0: No such file or directory
> ERROR: ../src/main.c:106: Failed to start IPTS: No such file or directory

It seems like the ipts kernel module is not loaded? Also the ipts kernel module needs to be loaded with some parameter set otherwise it won't work with gen 7+ devices. Can't remember what it is called (`gen7mt` or something similar). Running `modinfo ipts` should get you the name.

So first load the module `sudo modprobe ipts gen7mt=1` and then start iptsd.

**Comment by gurrgur** [comment]
> Not sure why it wouldn't work without poll in that case... @BeatVids @gurrgur Did you need the nosid or poll workarounds?

Yes, I needed the nosid workaround but it worked with both poll and interrupts.

**Comment by quo** [comment]
I see what happened with iptsd; you used the installation instructions in the readme, which are for the old version.

You can remove the iptsd directory you have now, then run the following commands to build and run iptsd:
```
git clone https://github.com/quo/iptsd
cd iptsd
meson build
ninja -C build
sudo ./build/src/daemon/iptsd
```

> Also the ipts kernel module needs to be loaded

No, you should not need the ipts module for SP8. Only ithc + iptsd.

> Yes, I needed the nosid workaround but it worked with both poll and interrupts.

Alright, yeah, then `sudo make set-nosid` is required on SP8 as well.

**Comment by luca-viggiani** [comment]
> I see what happened with iptsd; you used the installation instructions in the readme, which are for the old version.
> 
> You can remove the iptsd directory you have now, then run the following commands to build and run iptsd:
> 
> ```
> git clone https://github.com/quo/iptsd
> cd iptsd
> meson build
> ninja -C build
> sudo ./build/src/daemon/iptsd
> ```
> 
> > Also the ipts kernel module needs to be loaded
> 
> No, you should not need the ipts module for SP8. Only ithc + iptsd.
> 
> > Yes, I needed the nosid workaround but it worked with both poll and interrupts.
> 
> Alright, yeah, then `sudo make set-nosid` is required on SP8 as well.

Hi just a questiona about `sudo make set-nosid`. I see from the `README` that this is if you're using GRUB. In my case, as I'm using systemd boot, do I build it normally and then manually add the kernel param `intremap=nosid` and reboot, right?

**Comment by j4ckgrey** [comment]
I am going to clean install today again, and hope it works properly this time ... thanks for the follow up, much appreciated.

**Comment by quo** [comment]
Well, it looks like it worked?
If touch doesn't work, can you post `sudo dmesg` output?

**Comment by j4ckgrey** [comment]
> Well, it looks like it worked? If touch doesn't work, can you post `sudo dmesg` output?

I got touch to work withset-nosid, but iptsd for multi touch is not working ... I ran the commands you sent me, and it stops there

**Comment by quo** [comment]
@luca-viggiani 
> In my case, as I'm using systemd boot, do I build it normally and then manually add the kernel param `intremap=nosid` and reboot, right?

Yeah, that should work.

**Comment by j4ckgrey** [comment]
> It stops there because it's running. :) Multitouch should work while its running like that. If it doesn't, you can try stopping iptsd with Ctrl+C, then reloading the driver with `sudo modprobe -r ithc; sudo modprobe ithc`, then starting iptsd again.
> 
> If multitouch still doesn't work, can you do the following:
> 
>     * stop iptsd again
> 
>     * `sudo modprobe -r ithc; sudo modprobe ithc`
> 
>     * `sudo cat /dev/ithc > ithc.dat` (leave this running)
> 
>     * touch the screen a couple times
> 
>     * stop the above `cat` command with Ctrl+C
> 
>     * attach the ithc.dat file that was created (the file contains the raw touch data)

everything is working now, forgive my ignorance im no wizard, just a muggle here.
One question though, do I have to run iptsd everytime to have multi touch? cause once i stop it no multi touch

**Comment by j4ckgrey** [comment]
> @j4ckgrey Yes, you'll need IPTSd for multi-touch. We're essentially getting raw data that needs to be processed to be usable for the kernel. IPTSd does that and without it you'd have to rely on the built-in fallback mode, which requires no processing but is single-touch only.

Thanks, mate.
Thanks to all of you, it's a bless to have such community.

**Comment by j4ckgrey** [comment]
1 last thing though, I installed Kubuntu and once installed i checked the display settings and it had 120hz display available. Not sure if this can help but maybe we can get this to work on the kernel?

**Comment by k4ever** [comment]
When will the main installation instructions page be updated to include single touch, multi-touch, and pen support instructions for the Surface Pro 7+ and 8?

**Comment by qzed** [comment]
@k4ever Well, the plan is to include support in the kernel and IPTSd once it's matured a bit. So instructions don't have to be updated once we "officially" support it. At the moment it's all still work in progress.

**Comment by gurrgur** [comment]
I have done some research on this and for debugging it is very useful to enable dynamic debug messages for kms and the i915 driver. Apparently the 120Hz mode gets pruned because the driver sets up the 120Hz mode

**Comment by j4ckgrey** [comment]
> @j4ckgrey I think you just have to install pkg-config. On Ubuntu you should be able to install it with `sudo apt install pkg-config`.
> 
> As for poll mode, did you get the "Blocked an interrupt request" error in dmesg? If so, it's best to use the `intremap=nosid` workaround from the readme (`sudo make set-nosid` should work on Ubuntu).

hi there, i tried to search the answer before asking you but i failed. How to make set-nosid in archlinux?

**Comment by quo** [comment]
Depends on the bootloader you're using, see: https://wiki.archlinux.org/title/kernel_parameters

**Comment by j4ckgrey** [comment]
> Depends on the bootloader you're using, see: https://wiki.archlinux.org/title/kernel_parameters

I simply installed manjaro with efi partition

**Comment by quo** [comment]
From what I can tell `sudo make set-nosid` should work on Manjaro if using GRUB. Otherwise you'll probably have better luck asking on the Manjaro forums.

If you do get it working, feel free to post the solution here, I might be able to incorporate it into the Makefile.

**Comment by j4ckgrey** [comment]
> From what I can tell `sudo make set-nosid` should work on Manjaro if using GRUB. Otherwise you'll probably have better luck asking on the Manjaro forums.
> 
> If you do get it working, feel free to post the solution here, I might be able to incorporate it into the Makefile.

So I ran the command and I keep getting this error. Do you think Im missing something to install before running the command?

I only wanna make sure Im not missing anything before asking in their forums. Thanks in advance, mate.
![1 (2)](https://user-images.githubusercontent.com/47457650/172993004-9f2babdc-b62a-4c6f-8ef6-3e4ad54de1db.jpeg)

**Comment by NP-chaonay** [comment]
> Thanks, mate.

that sh shell doesnt allow stdout pipe output to the path that the subdirectory doesnt existed, you have to do mkdir -p follow with that path (exclude file just dir)

and then re run command,

pls make sure that your installed grub config to read config on that directory (generally should be)

**Comment by j4ckgrey** [comment]
> > Thanks, mate.
> 
> that sh shell doesnt allow stdout pipe output to the path that the subdirectory doesnt existed, you have to do mkdir -p follow with that path (exclude file just dir)
> 
> and then re run command,
> 
> pls make sure that your installed grub config to read config on that directory (generally should be)
> 
> @[j4ckgrey](https://github.com/j4ckgrey)

Thanks mate, I will try it right after work .... I wanted to do the same last night but as I have not enoug knowledge i wanted to avoid fucking up.

**Comment by j4ckgrey** [comment]
I created the mkdir and the command went through but touch didnt work, here is the output
![PXL_20220610_193140468](https://user-images.githubusercontent.com/47457650/173120763-ec61f26d-d36f-40ec-8530-8453198d9c18.jpg)

**Comment by j4ckgrey** [comment]
It only worked when i ran the command
sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dynbg=+pflmt

**Comment by NP-chaonay** [comment]
> It only worked when i ran the command sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dynbg=+pflmt

could you try to reboot to grub and see if kernel cmdline is updated, if not then your grub pkg of your distro is not config to read grub config from /etc/default/grub.d. if this happens then it up to implementation of your distro's grub, you have to figure where to set default kernel cmdline and save there

the another easy way is to put intremap=nosid in grub every boot manually

**Comment by j4ckgrey** [comment]
> > It only worked when i ran the command sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dynbg=+pflmt
> 
> could you try to reboot to grub and see if kernel cmdline is updated, if not then your grub pkg of your distro is not config to read grub config from /etc/default/grub.d. if this happens then it up to implementation of your distro's grub, you have to figure where to set default kernel cmdline and save there
> 
> the another easy way is to put intremap=nosid in grub every boot manually

Thanks, I think I will go for Ubuntu again .... Sad part is 120hz did not work there

**Comment by aleksfadini** [comment]
> @quo i didnt get touch working at first so I did enter poll mode (this command: sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt), and it worked only then in one a single touch mode (I already have Linux-Surface kernel installed), but I tried to install iptsd and I keep getting ERROR: Pkg-config binary for machine 1 not found. Giving up.
> 
> Is there something im doing wrong?

Just got a SP8, and 
`sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt`
worked to get my touchscreen working too!! That's amazing!!
Infinite thanks you guys are awesome.

**Comment by j4ckgrey** [comment]
> > @quo i didnt get touch working at first so I did enter poll mode (this command: sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt),
> 
> Thank you for this. Just got a SP8, and `sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt` worked to get my touchscreen working too (single touch)!! That's amazing!! Infinite thanks you guys are awesome.
> 
> On arch, it's just a matter of following the installation guide, then installing ithc after installing the dependencies with pacman, which on arch obviously are: `sudo pacman -S dkms linux-surface-headers` then install ithc as explained here https://github.com/quo/ithc-linux then reboot, then run `sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt`
> 
> Having single touch makes it quite tabletable for me!!

This command allows you to use touch only when you run, which means you have to run the command everytime you restart etc.
Run "sudo make set-nosid" to get touch working without having to do this everytime (works only in Ubuntu, couldnt get it to work in arch), and install iptsd experimental for multitouch, works fine as well.

**Comment by NP-chaonay** [comment]
> > > @quo i didnt get touch working at first so I did enter poll mode (this command: sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt),
> > 
> > 
> > Thank you for this. Just got a SP8, and `sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt` worked to get my touchscreen working too (single touch)!! That's amazing!! Infinite thanks you guys are awesome.
> > On arch, it's just a matter of following the installation guide, then installing ithc after installing the dependencies with pacman, which on arch obviously are: `sudo pacman -S dkms linux-surface-headers` then install ithc as explained here https://github.com/quo/ithc-linux then reboot, then run `sudo modprobe -r ithc; sudo modprobe ithc poll rx0 logregs dyndbg=+pflmt`
> > Having single touch makes it quite tabletable for me!!
> 
> This command allows you to use touch only when you run, which means you have to run the command everytime you restart etc. Run "sudo make set-nosid" to get touch working without having to do this everytime (works only in Ubuntu, couldnt get it to work in arch), and install iptsd experimental for multitouch, works fine as well.

idk if can using poll and the others permanently by set "ithc.poll=1 ithc.rx0 ithc.logregs ithc.dyndbg=+pflmt" in cmdline or if not able then maybe settings in modprobe (or distro's module loader setting).
intremap=nosid seem to be kernel cmdline and i cannot find where to set in /proc /sys so maybe require reboot.
while these ithc parameter can re-apply by reloading (some driver's parameter can change on run-time by change file in sysfs procfs)

**Comment by j4ckgrey** [comment]
> now that we have working multi-touch with iptsd and ithc are there any plans of incorporating them into this project and the repositories that this project runs?

I had the same thought today, would be much easier.

**Comment by aleksfadini** [comment]
I second that, I am also using both, and they work pretty well!

**Comment by qzed** [comment]
I don't have any particular objections against that, as long as @quo thinks it's ready and @StollD as doesn't have any issues with that either.

**Comment by j4ckgrey** [comment]
> I don't have any particular objections against that, as long as @quo thinks it's ready and @StollD as doesn't have any issues with that either.

I just wanna mention that there were still some troubles regarding archlinux for example, i couldnt get it installed as i remember then i was forced to go back to ubuntu. IPTSD doesnt work on Ubuntu 20.04 focal fossa. Same as the edid for 120hz ..... not sure if including them in the kernel would fix those things. Please enlighten me!

**Comment by StollD** [comment]
> I don't have any particular objections against that, as long as @quo thinks it's ready and @StollD as doesn't have any issues with that either.

If the kernel driver is ready we should include it.

Regarding iptsd, I had hoped that I would have the time at some point to dive into all of the HID stuff, so that we could adapt iptsd to use the HID interface on all platforms (GuC based IPTS, IPTS on gen7 and ITHC). I also wrote some code for it but ran into some issues with the data format, where I didnt have the time to think about the best solution yet.

**Comment by quo** [comment]
I was hoping to find a better fix for the irq source id issue, so the `intremap=nosid` workaround wouldn't be necessary. The best I have been able to come up with is a patch to disable the sid check for ITHC only instead of globally (see https://github.com/quo/ithc-linux/issues/2).

Since other than that the driver seems to work fine, I'm ok with including it in the surface kernel. You'll have to add the above patch as well.

As for iptsd, feel free to merge my changes. Unfortunately I haven't had time to test the parser changes on older SP models yet (but if anything's broken it should be trivial to fix).

I still want to add tilt support in the short term (should be fairly straightforward, see https://github.com/quo/iptsd/issues/3), and also make the pen signal detection / position fitting a little more robust, but I just haven't had the time.

**Comment by StollD** [comment]
I took a look at your changes, I actually just made some very similar ones on the hid branch of iptsd. The DFT stuff seems simple enough to port. Once I get the whole daemon working with the HID based driver, I can either try to port it myself, or ping you so you can make a pull request. Whatever you prefer.

Other than that I hope the code will just work (tm) with the ITHC driver in HID mode. Since you say in the README that the mode is not recommended: Is that just because there is no HID-enabled iptsd, or is there anything else that doesnt work well in that mode?

**Comment by quo** [comment]
Would be great if you could port it. Feel free to use the parts you like and ignore the rest :) (and don't worry about credit or anything, I'm happy as long as my code gets used somewhere).

I do recommend adding something like the Block abstraction I made; it makes the parsing code easier to read and is a little less error-prone than manually tracking sizes and seeking/skipping through the data.

And yeah, the main reason the readme says HID is not recommended is that it has not really been tested sufficiently. The ithc driver currently does enable HID by default if the touch device is not an MS device, because on the Lenovo X1 Fold the THC is connected to Wacom hardware which sends normal HID reports.

**Comment by StollD** [comment]
I pushed a port of @quo's code into the `dft` branch on iptsd. I would be interested to hear if that branch

* works on gen7 and above
* behaves like (or at least similar) to the iptsd from @quo 

If you have a Surface 7 Pro, Surface Laptop 3 or Surface Book 3 you will need to build the ipts module from the `hid` branch: https://github.com/linux-surface/intel-precise-touch/tree/hid

```
$ git clone https://github.com/linux-surface/intel-precise-touch
$ cd intel-precise-touch
$ git checkout hid
$ make
$ sudo systemctl stop iptsd
$ sudo rmmod ipts
$ sudo insmod ipts.ko
```

If you have a Surface Pro 7+ or later, you will need to use ITHC from @quo in HID mode: 

```
$ git clone https://github.com/quo/ithc-linux/
$ cd ithc-linux
$ make
$ sudo insmod build/ithc.ko hid=1
```

**Comment by StollD** [comment]
You only need to run one of the sections, depending on what device you have. SP8 uses THC, so you need ITHC.

Since you already have a working setup, all you really need to do regarding the kernel is add `ithc.hid=1` to the kernel commandline, and then build and launch iptsd from the `dft` branch.

**Comment by StollD** [comment]
There are also prebuilt packages for the distributions we support: https://github.com/linux-surface/iptsd/actions/runs/2699184935 

Download the artifact called `<your distro>-latest`, unpack the zip file and install the package inside. Then reboot, and load the kernel module (if not done automatically with DKMS etc.). The service should start automatically and multitouch should work.

You can run this to check if the service was automatically started
```
ps aux | grep iptsd
```

**Comment by aleksfadini** [comment]
> dmesg | grep -i hidraw | grep -E '(IPTS|Intel Touch Host Controller)'

I'm in the process of compiling iptsd from the dft branch. The command above doesn't give any result. I'm looking for my hidraw number. I have 0 to 4.

```
# dmesg | grep -i hidraw | grep -E '(IPTS|Intel Touch Host Controller)'
[root@AgileFenix aleks]# dmesg | grep -i hidraw
[    2.256346] hid-generic 0019:045E:0922.0001: hidraw0: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0922] on 01:15:01:07:00
[    2.291113] hid-generic 0019:045E:0922.0002: hidraw1: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0922] on 01:15:01:06:00
[    2.456272] hid-generic 0019:045E:09AE.0003: input,hidraw1: <UNKNOWN> HID v1.11 Keyboard [Microsoft Surface 045E:09AE] on 01:15:02:01:00
[    2.462163] hid-generic 0019:045E:0991.0004: hidraw2: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0991] on 01:15:02:05:00
[    2.466238] hid-generic 0019:045E:09B0.0005: input,hidraw3: <UNKNOWN> HID v1.11 Keyboard [Microsoft Surface 045E:09B0] on 01:15:02:02:00
[    2.470312] hid-generic 0019:045E:09AF.0006: input,hidraw4: <UNKNOWN> HID v1.11 Mouse [Microsoft Surface 045E:09AF] on 01:15:02:03:00
[    2.682376] hid-multitouch 0019:045E:09AF.0006: input,hidraw4: <UNKNOWN> HID v1.11 Mouse [Microsoft Surface 045E:09AF] on 01:15:02:03:00
```

Which one is it?

**Comment by qzed** [comment]
Those look all like Surface Aggregator devices (on account of the `01:15:...`). Is the `ithc` module loaded and in `hid` mode?

**Comment by aleksfadini** [comment]
Sorry, I think the ithc module is loaded but not in hid mode. I'm compling this while using the latest SP8 fixes, which included the iptsd from @quo and no hid mode (I suppose). How would I check?

**Comment by qzed** [comment]
As far as I can tell you need to specify `hid=1` when loading the module, e.g. `sudo insmod build/ithc.ko hid=1` as mentioned by @StollD above.

**Comment by StollD** [comment]
If you have it installed via. DKMS, you can add `ithc.hid=1` to the kernel commandline.

**Comment by StollD** [comment]
I pushed the configs for SP7+ and SP8 to the dft branch. Running `git pull` should fix the display size error.

**Comment by aleksfadini** [comment]
I'm still getting the display size error after a git pull. I even deleted the folder and re-cloned and rebuilt (from dft) to make sure.

**Comment by StollD** [comment]
> Are we sure 4 is my hidraw device?

We are sure that it isnt because your ithc is not in HID mode. Since hidraw4 is a different device, the device ID will not be the one thats in the config I pushed, so it still wont find one and error out.

**Comment by aleksfadini** [comment]
I am not sure what I'm missing.
I have installed ithc via dkms, as explained earlier in this thread.
In my GRUB I have:

```
GRUB_CMDLINE_LINUX_DEFAULT="loglevel=3 quiet i915.enable_psr=0 ithc.hid=1"
```
So I'm expecting HID to be on, yet I still get:


```
sudo dmesg | grep -i hidraw
[    2.202810] hid-generic 0019:045E:0922.0001: hidraw0: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0922] on 01:15:01:07:00
[    2.236413] hid-generic 0019:045E:0922.0002: hidraw1: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0922] on 01:15:01:06:00
[    2.474442] hid-generic 0019:045E:09AE.0003: input,hidraw1: <UNKNOWN> HID v1.11 Keyboard [Microsoft Surface 045E:09AE] on 01:15:02:01:00
[    2.479415] hid-generic 0019:045E:0991.0004: hidraw2: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0991] on 01:15:02:05:00
[    2.484620] hid-generic 0019:045E:09B0.0005: input,hidraw3: <UNKNOWN> HID v1.11 Keyboard [Microsoft Surface 045E:09B0] on 01:15:02:02:00
[    2.488598] hid-generic 0019:045E:09AF.0006: input,hidraw4: <UNKNOWN> HID v1.11 Mouse [Microsoft Surface 045E:09AF] on 01:15:02:03:00
[    2.614290] hid-multitouch 0019:045E:09AF.0006: input,hidraw4: <UNKNOWN> HID v1.11 Mouse [Microsoft Surface 045E:09AF] on 01:15:02:03:00
```

So I don't know which hidraw to use, and  still get the "display zero" error.

**Comment by StollD** [comment]
Can you post the entire output of `dmesg` please?

**Comment by qzed** [comment]
Those are still all Surface Aggregator HID devices. You're looking for something that doesn't have the `on 01:15:...` at the end.

And it looks like `ithc` doesn't want to load:
```
[    2.693583] ithc 0000:00:10.6: ithc_probe: hid_add_device failed with -110
[    2.693586] ithc: probe of 0000:00:10.6 failed with error -110
```

Error -110 is ETIMEDOUT. Unfortunately I don't know anything about the ithc driver, so I can't say why this happens. I guess @quo would know more.

**Comment by qzed** [comment]
Do you still have `intermap=nosid` set?

**Comment by aleksfadini** [comment]
> Do you still have `intermap=nosid` set?

Ahhh! No I do not, because I thought I needed to take it out to test. Good point, will put it back and try again!

**Comment by aleksfadini** [comment]
> Is the iptsd from quo you were running before still active by any chance?

It might indeed be! What would be a way to check on that?

**Comment by qzed** [comment]
Assuming you've started it via the systemd service, `systemctl status iptsd.service`. To stop it run `systemctl stop iptsd.service`.

**Comment by aleksfadini** [comment]
Thank you, I understand this was obvious in hindsight. 
After disabling the ipsd service, the functionality is still buggy, but in a slightly different way. I tested extensively, and the pinch motions are particularly buggy (ex. if you pinch 5 times in a row, with the same motion, and pause between pinching, number 3-5 will not work. If you wait longer, pinch 6 will work).
.
I'd say the @quo iptsd was more stable for multi-touch, but this branch works better for the pen.

For some reason the SP8 Pen now works perfectly, including the back button (eraser). The middle button does nothing.
On the other hand, with the @quo iptsd, the pen was buggy (sometimes missing lines) but the middle button was detected, although it would work a bit randomly.

I can make a small video if needed (let me know)

**Comment by StollD** [comment]
The only touch related difference between the two iptsd's should be this change, that I backported from the C version master: https://github.com/linux-surface/iptsd/commit/f20c82b6d8e8acd21ba6b146b424540212e2a771

Could you try reverting the change, and see if that changes the touch behaviour? Run the following command in the iptsd source dir and then rebuild.
```
$ git revert f20c82b6d8e8acd21ba6b146b424540212e2a771
```

**Comment by aleksfadini** [comment]
Thank you very much for digging deeper into this. Just to make sure I am not missing anything, after using the git revert command, when rebuilding I get this:

```
$ meson build --wrap-mode=forcefallback
Directory already configured.

Just run your build command (e.g. ninja) and Meson will regenerate as necessary.
If ninja fails, run "ninja reconfigure" or "meson --reconfigure"
to force Meson to regenerate.

If build failures persist, run "meson setup --wipe" to rebuild from scratch
using the same options as passed when configuring the build.
To change option values, run "meson configure" instead.
[aleks@AgileFenix iptsd]$ ninja -C build
ninja: Entering directory `build'
[8/8] Linking target src/daemon/iptsd
```
Was it correctly rebuilt, including the revert changes? Or should I have deleted the build directory entirely after git revert?

Because the same bugginess seems to happen.

**Comment by StollD** [comment]
> I tested more extensively, I do think that after the revert, the driver works exactly as it did with the @quo ipts.

That is good to know, thank you.

> On the other hand, the version _before_ git revert had much better pen support

To be honest, I have no idea why that would happen, giving the reverted commit does not touch anything related to the stylus. Could this be just random differences in behaviour between two test runs?

@qzed found that on Surface Pro X, changing this value from 1000 to 4000 yielded some improvements. Maybe you could try that? https://github.com/linux-surface/iptsd/blob/dft/src/daemon/dft.cpp#L17

> https://youtu.be/96ToS8E6m1s

The pinch / zoom issue looks a bit like palm detection is kicking in, marking one contact as a palm. As a safety measure, contacts that were marked as a palm once will stay a palm forever, until they are lifted. This might simply come down to tweaking the thresholds a bit. Make sure that you are not pressing too hard, otherwise it could confuse your finger for a palm.

**Comment by StollD** [comment]
> The pinch / zoom issue looks a bit like palm detection is kicking in, marking one contact as a palm. As a safety measure, contacts that were marked as a palm once will stay a palm forever, until they are lifted. This might simply come down to tweaking the thresholds a bit. Make sure that you are not pressing too hard, otherwise it could confuse your finger for a palm.

I think this might be what is causing your initial issues as well (before the revert of that commit). Its kinda hard to figure out remotely, but I dont have any issues with that commit (and the touch processing between SB2 and SP8 should be identical).

What you could try is switching to the advanced (or "qzed") touch processor. Maybe that can handle your fingers better :P Put this into `/etc/ipts.conf`

```
[Touch]
Processing = advanced
```

**Comment by j4ckgrey** [comment]
Is it still happening? are the drivers gonna get included in the Surface Kernal? If so, will the 120hz Edid be included too?

**Comment by StollD** [comment]
The plan is to merge the drivers into the surface kernel once everything is ready, yes. 

Bit offtopic for this, but for the EDID, I dont like the idea of including EDID hacks in the kernel because I think that could get messy and annoying to maintain. I vaguely remember that someone said it works fine on Ubuntu default kernels, which would mean the issue is probably caused by an upstream regression and should be reported / fixed there.

**Comment by zed-wong** [comment]
Hi quo, I'm using SP8 running Ubuntu 22.04 and I've done all these:
```
git clone https://github.com/quo/iptsd
cd iptsd
meson build
ninja -C build
sudo ./build/src/daemon/iptsd
```
It all went well. The last step shows:
```
[01:37:18.107] [info] Vendor/product: 045E:0C37
[01:37:18.111] [info] Opened /dev/ithc
```

But the touch screen still doesn't work. So I tried `make set-nosid` and it still doesn't work.
```
$ sudo make set-nosid
make: *** No rule to make target 'set-nosid'.  Stop.
```

How can I make it work? Thanks!
```
- Neofetch
OS: Ubuntu 22.04 LTS x86_64
Host: Surface Pro 8 
Kernel: 5.18.12-surface
```

**Comment by StollD** [comment]
Are you running that command in the iptsd directory or the ithc-linux directory? You need to run it in the ithc-linux directory.

**Comment by zed-wong** [comment]
> Are you running that command in the iptsd directory or the ithc-linux directory? You need to run it in the ithc-linux directory.

Thank you! I ran it in the ithc-linux directory and it succeed.
```
echo 'GRUB_CMDLINE_LINUX_DEFAULT="$GRUB_CMDLINE_LINUX_DEFAULT intremap=nosid"' > /etc/default/grub.d/intremap-nosid.cfg
update-grub
Sourcing file `/etc/default/grub'
Sourcing file `/etc/default/grub.d/init-select.cfg'
Sourcing file `/etc/default/grub.d/intremap-nosid.cfg'
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-5.18.12-surface
Found initrd image: /boot/initrd.img-5.18.12-surface
Found linux image: /boot/vmlinuz-5.16.17-surface
Found initrd image: /boot/initrd.img-5.16.17-surface
Found linux image: /boot/vmlinuz-5.15.6-surface
Found initrd image: /boot/initrd.img-5.15.6-surface
Found linux image: /boot/vmlinuz-5.15.0-41-generic
Found initrd image: /boot/initrd.img-5.15.0-41-generic
Memtest86+ needs a 16-bit boot, that is not available on EFI, exiting
Warning: os-prober will be executed to detect other bootable partitions.
Its output will be used to detect bootable binaries on them and create new boot entries.
Found Windows Boot Manager on /dev/nvme0n1p1@/EFI/Microsoft/Boot/bootmgfw.efi
Adding boot menu entry for UEFI Firmware Settings ...
done
sync
```

But the touchscreen is still not working. I'm gonna try to reboot to see if it works.

**Comment by zed-wong** [comment]
It's working now! But the multi-touch is not working yet.

**Comment by zed-wong** [comment]
> It's working now! But the multi-touch is not working yet.

I tried the method quo mentioned above. The multi-touch works. I tested `three fingers drag up/down/left/right` and those are working properly. But `two fringers drag up/down`, `two fingers click` and `single long press` didn't work.

**Comment by StollD** [comment]
iptsd does not do any gesture detection. It merely determines the positions of all fingers on the display and sends that to userspace. The userspace will then do its own gesture detection.

GNOME Shell for example doesn't have any two finger gestures, only three finger ones. If you want to try a two finger gesture, try zooming in Firefox.

**Comment by zed-wong** [comment]
> iptsd does not do any gesture detection. It merely determines the positions of all fingers on the display and sends that to userspace. The userspace will then do its own gesture detection.
> 
> GNOME Shell for example doesn't have any two finger gestures, only three finger ones. If you want to try a two finger gesture, try zooming in Firefox.

Zooming in Firefox doesn't work. It also doesn't work with my touchpad though, have to press on ctrl.

**Comment by StollD** [comment]
> Zooming in Firefox doesn't work. It also doesn't work with my touchpad though, have to press on ctrl.

Are you running a Wayland session? If yes, is Firefox also running in Wayland mode? Run `xprop` and click in the Firefox window. If there is an output in the console, its running using X11. The touch / gesture support of X11 is kinda bad because every input emulates a mouse pointer.

**Comment by zed-wong** [comment]
Yes, it's Wayland.
![image](https://user-images.githubusercontent.com/36271125/181439552-122137d4-9d7a-49f5-8522-8c303abdcca0.png)
And `xprop` has output:
![image](https://user-images.githubusercontent.com/36271125/181439615-89c88a92-c8eb-4a14-8ccd-76adfa83f25f.png)

**Comment by StollD** [comment]
Then try with `env MOZ_ENABLE_WAYLAND=1 firefox`

**Comment by zed-wong** [comment]
> Then try with `env MOZ_ENABLE_WAYLAND=1 firefox`

Two finger gesture is working!

**Comment by xrishox** [comment]
are there any issues left that people are running into?

**Comment by j4ckgrey** [comment]
Actually i was wondering when the drivers are getting merged in the kernel? Everything seems to be fine.

**Comment by StollD** [comment]
> Actually i was wondering when the drivers are getting merged in the kernel? Everything seems to be fine.

Most likely when iptsd is ready.

**Comment by xrishox** [comment]
maybe a bit of a weird question. once all of this stuff is merged will dkms still be required to load the modules or will they be fully built into the kernel? i would like to ultimately land on silverblue, but silverblue doesn't actually support dkms.

**Comment by qzed** [comment]
@xrishox Once included, the kernel modules will be fully built in, so no dkms required.

**Comment by JoshuaFarrow** [comment]
Hi guys, I've managed to get ITHC running and the touch is working fine apart from dual touch. I've ran `sudo sysytemctl status iptsd` and I am getting  an error:

```
systemd[1]: Started Intel Precise Touch & Stylus Daemon.
iptsd[5618]: ERROR: ../src/control.c:139: Failed to open /dev/ipts/0: No such file or directory
iptsd[5618]: ERROR: ../src/main.c:106: Failed to start IPTS: No such file or directorysystemd[1]: iptsd.service: Main process exited, code=exited, status=254/n/a
systemd[1]: iptsd.service: Failed with result 'exit-code'.
```
Does anybody know what might be causing this?

**Comment by NP-chaonay** [comment]
> 

try to using iptsd from linux-surface (recompile or prebuilt from Github only at this moment), it is now available for usage with ithc.

**Comment by JoshuaFarrow** [comment]
@NP-chaonay 

Thanks NP, I've tried building and running with ninja on the instructions but I'm getting a few errors.:

`ERROR: Command "/home/josh/Downloads/iptsd/subprojects/hidrd/bootstrap" failed with status 127`

**Comment by surli812** [comment]
@JoshuaFarrow did you try using one of the pre-built packages? You'll need to be signed-in to Github, then go to https://github.com/linux-surface/iptsd/actions, select the last successful build and download the artifact named <your distro>-latest

**Comment by mityax** [comment]
Hi, since this thread is very long and I'm not sure what's up to date and what's not anymore:

What steps would be required for setting up touch and stylus support on a fresh (fedore37) Installation in SP8?
Thanks for and help :)

**Comment by xrishox** [comment]
any news on getting this merged?

**Comment by aleksfadini** [comment]
@mityax  I don't even remember what steps I took on my end, 5 months ago, and I have everything working now. It's buried at the top of this and other threads. Would be nice to have a clean recap for future users while we wait for everything to be merged.

**Comment by mityax** [comment]
@aleksfadini yes, that'd be really nice. I cleaned up and updated my comment above, and also published the script I use after kernel updates (for some reason, I need to redo most of the steps in this scenario). Hope it helps someone.

**Comment by mr-rodgers** [comment]
> [...] did you try using one of the pre-built packages? You'll need to be signed-in to Github, then go to https://github.com/linux-surface/iptsd/actions, select the last successful build and download the artifact named -latest

Not OP, but when I install the .deb package from Actions (Ubuntu 22.10), the systemctl config for iptsd breaks, and then touch doesn't work anymore

I also cant build from source, because I cant get the hidrd dep installed. 

Is there something else I can try?

**Comment by StollD** [comment]
> Not OP, but when I install the .deb package from Actions (Ubuntu 22.10), the systemctl config for iptsd breaks, and then touch doesn't work anymore

What error are you experiencing? You are not supposed to run any systemctl commands with the new iptsd, it will be launched automatically by udev when it detects a HID device from a supported kernel driver (ITHC, IPTS built from source and SPI-HID).

**Comment by mr-rodgers** [comment]
> What error are you experiencing? You are not supposed to run any systemctl commands with the new iptsd, it will be launched automatically by udev when it detects a HID device from a supported kernel driver (ITHC, IPTS built from source and SPI-HID).

I switched to fedora 37, but it's the same error. I install the .rpm using `sudo dnf localinstall ./downloaded.rpm`

Then I get this when I try to check the status of iptsd

```
[te-je@fedora fedora-37-latest]$ sudo systemctl status iptsd
[sudo] password for te-je: 
● iptsd.service
     Loaded: not-found (Reason: Unit iptsd.service not found.)
     Active: active (running) since Fri 2022-12-30 20:13:15 CET; 39min left
   Main PID: 830 (iptsd)
        CPU: 291ms
     CGroup: /system.slice/iptsd.service
             └─830 /usr/bin/iptsd

Dec 30 20:13:15 fedora systemd[1]: Started iptsd.service - Intel Precise Touch & Stylus Daemon.
Dec 30 19:26:54 fedora systemd[1]: iptsd.service: Current command vanished from the unit file, execution of the command list won't be resumed.
``

**Comment by StollD** [comment]
I honestly dont even know why you still have a `iptsd.service`. The iptsd from actions uses a parameterized service file (so `iptsd@.service`). To get the status of iptsd run `systemctl status iptsd@$(sudo iptsd-find-hidraw).service`

**Comment by mr-rodgers** [comment]
The command `sudo iptsd-find-hidraw` returns nothing at all. I think maybe this should not be?

**Comment by StollD** [comment]
What device do you have? You definitly have to install a kernel driver with DKMS, depending on the device thats either IPTS (Until "gen7" aka. SP7, SL3, SB3) or ITHC (SP7+ and later).

**Comment by mr-rodgers** [comment]
Is there some docs on how to install the kernel driver?

I am on SP7.

**Comment by StollD** [comment]
https://github.com/linux-surface/intel-precise-touch#building-dkms

Ignore the first bullet point (applying patches), the surface kernel already includes them.

You also need to disable secureboot so that the unsigned DKMS module can be loaded.

**Comment by mr-rodgers** [comment]
Restarting after installing the dkms driver from the repo bricked the wifi and bluetooth on the device (they no longer appear).

And after checking uname -a I can see that I got booted into the linux kernel, and not the linux-surface kernel. Do you have any idea what happened there?

**Comment by StollD** [comment]
You updated the kernel and now there is a kernel from Fedora that is newer than the surface kernel. We have a fix for that here: https://github.com/linux-surface/linux-surface/tree/master/contrib/fedora-default-kernel

**Comment by mr-rodgers** [comment]
Hey thanks a lot for helping me with this! I got the multitouch working now on SP7

**Comment by Ngin77** [comment]
> `sudo modprobe ithc`

Hi. Unfortunaley I get below error, if I try to enter modrpobe command at point 12.
![Auswahl_001](https://github.com/linux-surface/linux-surface/assets/11647930/28353b0b-e50e-44ed-b361-ea91df323ba1)
Greetings.

**Comment by mityax** [comment]
@Ngin77 i have no idea about the error you get, sorry. However, for quite some months now it has been enough for me to just install using the normal procedures/package repos. I think they upstreamed some changes or so and now there's no need for these custom steps anymore - maybe you should try this if you haven't already

**Comment by StollD** [comment]
Yes, the latest surface kernel builds already include the latest version of the ITHC driver.

However, you must uninstall the DKMS version before installing the latest kernel, because otherwise the DKMS build will replace the one shipped with the surface kernel.

In your specific case, the DKMS build is not signed with a trusted key, and the Fedora kernel will refuse to load untrusted modules when secureboot is enabled.

**Comment by Ngin77** [comment]
> @Ngin77 i have no idea about the error you get, sorry. However, for quite some months now it has been enough for me to just install using the normal procedures/package repos. I think they upstreamed some changes or so and now there's no need for these custom steps anymore - maybe you should try this if you haven't already

@mityax THX a lot. May be you have a direct link for me or closer information? Would be great. I already searched after any installation guides for SF8 and touch panel usage, but didn't find something, but this side / tutorial.

@StollD So should I uninstall and disable secure boot? Because in my case I have been turned off secure boot for dual boot (windows and Fedora).

**Comment by StollD** [comment]
Just follow the normal installation guide. Nothing special about SP8.

You should uninstall the DKMS module. After that you can either disable secureboot, or keep it enabled (if the key that we use for signing the surface kernel is enrolled, but that is also covered by the installation guide).

**Comment by Ngin77** [comment]
So I upgraded from version 37 to 38, but the touchscreen still doesn't work ....
Any hints pls?
Sincerely.

**Comment by StollD** [comment]
That is not exactly a very informative report.

Make sure the kernel driver is loaded and you have the latest iptsd installed. If you need any more help, you need to provide at least a dmesg.

**Comment by Ngin77** [comment]
> That is not exactly a very informative report.
> 
> Make sure the kernel driver is loaded and you have the latest iptsd installed. If you need any more help, you need to provide at least a dmesg.

Hi and thank you.

Here there are dmesg and lsmod log entries. In dmesg (line 1002) there is closer information to the touchpanel. But I don't know, which driver I need exactly
[dmesg.txt](https://github.com/linux-surface/linux-surface/files/11662731/dmesg.txt)
[lsmod.txt](https://github.com/linux-surface/linux-surface/files/11662733/lsmod.txt)

Thank you in advance.
Best regards.

**Comment by StollD** [comment]
You still have the old unsigned DKMS module of ITHC installed. I told you you need to remove it. In the ithc repository that you cloned to install it, run `sudo make dkms-uninstall`

Since the DKMS module is not signed it cannot load and iptsd does not start.

**Comment by Ngin77** [comment]
@StollD Thank you. Cannot remove dkms in repository unfortunately:

dkms status:

ithc/0.1, 6.3.5-100.fc37.x86_64, x86_64: installed
ithc/0.1, 6.3.5-200.fc38.x86_64, x86_64: installed

sudo make dkms-uninstall:

make: *** No rule to build dkms-uninstall. Ending. (translated from german)

Thank you again.

Sincerely.
